### PR TITLE
feature/pagination

### PR DESCRIPTION
**Pull request by dovgopoly**

### CODE REVIEW

File: contracts/gov/validators/GovValidators.sol [hunk 1]
```diff
@@ -19,11 +19,11 @@ contract GovValidators is IGovValidators, OwnableUpgradeable {
...
     InternalProposalSettings public internalProposalSettings;
 
-    uint256 internal _latestInternalProposalId;
+    uint256 public override latestInternalProposalId;
     uint256 public validatorsCount;
 
-    mapping(uint256 => InternalProposal) public internalProposals; // proposalId => info
-    mapping(uint256 => ExternalProposal) public externalProposals; // proposalId => info
+    mapping(uint256 => InternalProposal) public _internalProposals; // proposalId => info
+    mapping(uint256 => ExternalProposal) public _externalProposals; // proposalId => info
```
Comment (Arvolear) [inline contracts/gov/validators/GovValidators.sol:26]: `internal` instead of `public`?

Comment (dovgopoly) [reply contracts/gov/validators/GovValidators.sol]: correct, fixed

File: contracts/interfaces/gov/IGovPool.sol [hunk 2]
```diff
@@ -155,4 +163,9 @@ interface IGovPool {
         external
         view
         returns (uint256, ShrinkableArray.UintArray memory);
+
+    function getProposals(uint256 offset, uint256 limit)
```
Comment (Arvolear) [inline contracts/interfaces/gov/IGovPool.sol:167]: Let's have the order of functions in the interface match the order in the contract

File: contracts/libs/gov-pool/GovPoolView.sol [hunk 3]
```diff
@@ -133,4 +136,24 @@ library GovPoolView {
...
+        IGovPool govPool = IGovPool(address(this));
+
+        uint256 to = (offset + limit).min(govPool.latestProposalId()).max(offset);
+
+        (, , address validators, ) = govPool.getHelperContracts();
+
+        proposalViews = new IGovPool.ProposalView[](to - offset);
+        for (uint256 i = offset; i < to; ++i) {
+            proposalViews[i - offset] = IGovPool.ProposalView({
+                proposal: proposals[i],
```
Comment (Arvolear) [inline contracts/libs/gov-pool/GovPoolView.sol:154]: Probably we should add `1` to `i` in order to fetch the correct proposal. They start from the index `1`, but `i` starts from `0`

File: contracts/gov/validators/GovValidators.sol [hunk 4]
```diff
@@ -175,6 +175,29 @@ contract GovValidators is IGovValidators, OwnableUpgradeable {
...
+        external
+        view
+        override
+        returns (InternalProposal[] memory internalProposals)
+    {
+        uint256 to = (offset + limit).min(latestInternalProposalId).max(offset);
+
+        internalProposals = new InternalProposal[](to - offset);
+        for (uint256 i = offset; i < to; ++i) {
+            internalProposals[i - offset] = _internalProposals[i];
```
Comment (Arvolear) [inline contracts/gov/validators/GovValidators.sol:197]: Same index issue here

File: contracts/interfaces/gov/IGovPool.sol (hunk 2 above)
Comment (dovgopoly) [reply contracts/interfaces/gov/IGovPool.sol]: done

File: contracts/gov/validators/GovValidators.sol (hunk 4 above)
Comment (dovgopoly) [reply contracts/gov/validators/GovValidators.sol]: fixed

File: test/GovPool.test.js [hunk 5]
```diff
@@ -388,7 +390,7 @@ describe("GovPool", () => {
       it("should create 2 proposals", async () => {
```
Comment (Arvolear) [inline test/GovPool.test.js:390]: Here, for example

File: test/GovValidators.test.js [hunk 6]
```diff
@@ -359,6 +361,94 @@ describe("GovValidators", () => {
       });
     });
 
+    describe("getExternalProposal()", () => {
+      it("should return zero proposal if doesnt exist", async () => {
```
Comment (Arvolear) [inline test/GovValidators.test.js:365]: Doesn't typo

**Review by Arvolear — APPROVED** [review]
Ok

**Pull request merged**